Source/Core/Core/HW/DVD/DVDInterface.h

```diff
@@ -97,12 +116,144 @@ bool UpdateRunningGameMetadata(std::optional<u64> title_id = {});
 
 // Direct access to DI for IOS HLE (simpler to implement than how real IOS accesses DI,
 // and lets us skip encrypting/decrypting in some cases)
-void ChangePartition(const DiscIO::Partition& partition);
 void ExecuteCommand(u32 command_0, u32 command_1, u32 command_2, u32 output_address,
                     u32 output_length, bool reply_to_ios);
+void PerformDecryptingRead(u32 position, u32 length, u32 output_address,
+                           const DiscIO::Partition& partition, bool reply_to_ios);
+
+void SetLowError(u32 low_error);
+void SetHighError(u32 high_error);
 
 // Used by DVDThread
 void FinishExecutingCommand(ReplyType reply_type, DIInterruptType interrupt_type, s64 cycles_late,
                             const std::vector<u8>& data = std::vector<u8>());
 
+// Hardware registers
+// DI Status Register
+union UDISR
+{
+  u32 Hex;
+  struct
+  {
+    u32 BREAK : 1;       // Stop the Device + Interrupt
+    u32 DEINITMASK : 1;  // Access Device Error Int Mask
+    u32 DEINT : 1;       // Access Device Error Int
+    u32 TCINTMASK : 1;   // Transfer Complete Int Mask
+    u32 TCINT : 1;       // Transfer Complete Int
+    u32 BRKINTMASK : 1;
+    u32 BRKINT : 1;  // w 1: clear brkint
+    u32 : 25;
+  };
+  UDISR() { Hex = 0; }
+  UDISR(u32 _hex) { Hex = _hex; }
+};
+
+// DI Cover Register
+union UDICVR
+{
+  u32 Hex;
+  struct
+  {
+    u32 CVR : 1;         // 0: Cover closed  1: Cover open
+    u32 CVRINTMASK : 1;  // 1: Interrupt enabled
+    u32 CVRINT : 1;      // r 1: Interrupt requested w 1: Interrupt clear
+    u32 : 29;
+  };
+  UDICVR() { Hex = 0; }
+  UDICVR(u32 _hex) { Hex = _hex; }
+};
+
+union UDICMDBUF
+{
+  u32 Hex;
+  struct
+  {
+    u8 CMDBYTE3;
+    u8 CMDBYTE2;
+    u8 CMDBYTE1;
+    u8 CMDBYTE0;
+  };
+};
+
+// DI DMA Address Register
+union UDIMAR
+{
+  u32 Hex;
+  struct
+  {
+    u32 Zerobits : 5;  // Must be zero (32byte aligned)
+    u32 : 27;
+  };
+  struct
+  {
+    u32 Address : 26;
+    u32 : 6;
+  };
+};
+
+// DI DMA Address Length Register
+union UDILENGTH
+{
+  u32 Hex;
+  struct
+  {
+    u32 Zerobits : 5;  // Must be zero (32byte aligned)
+    u32 : 27;
+  };
+  struct
+  {
+    u32 Length : 26;
+    u32 : 6;
+  };
+};
+
+// DI DMA Control Register
+union UDICR
+{
+  u32 Hex;
+  struct
+  {
+    u32 TSTART : 1;  // w:1 start   r:0 ready
+    u32 DMA : 1;     // 1: DMA Mode    0: Immediate Mode (can only do Access Register Command)
+    u32 RW : 1;      // 0: Read Command (DVD to Memory)  1: Write Command (Memory to DVD)
+    u32 : 29;
+  };
+};
+
+union UDIIMMBUF
+{
+  u32 Hex;
+  struct
+  {
+    u8 REGVAL3;
+    u8 REGVAL2;
+    u8 REGVAL1;
+    u8 REGVAL0;
+  };
+};
+
+// DI Config Register
+union UDICFG
+{
+  u32 Hex;
+  struct
+  {
+    u32 CONFIG : 8;
+    u32 : 24;
+  };
+  UDICFG() { Hex = 0; }
+  UDICFG(u32 _hex) { Hex = _hex; }
+};
+
+extern UDISR s_DISR;
```

Comment:
These should probably not all be exposed, but rather have an interface built around them.

Reply:
We already have these registers set up for MMIO (so that GameCube games can use them). Would it be cleaner to make IOS write to the emulated console's memory (triggering the MMIO handlers) instead of it doing what it does now? That way, it's even closer to the original hardware, and we don't have to make as many changes in what we expose.

Reply:
That would be ideal, actually.  Does that work?  I assumed `Memory::Write_U32` would only write to the main game memory and bypass MMIOs.

Reply:
I believe `Memory::Write_U32` works for that, yes.

Reply:
I've started implementing something like that and have changed to writing the DICMDBUF values at least.  `Memory::Write_U32` didn't work, but `Memory::mmio_mapping->Write` does.  I'm still trying to figure out how to handle the other values; reading a value and writing it back is a bit awkward (especially given the r/z nature of some of the registers &mdash; IOS can do `DISR = DISR & ~(DEINT|TCINT|BRKINT) | DEINTMASK` but it'd be a lot uglier to do the same with the current struct).  Maybe a hybrid approach is easiest, where the bitfields do actually have dedicated functions in DVDInterface instead...

Reply:
I've implemented it; the only things that needed special access were the interrupts (which already had a bit of abstraction via `DIInterruptType`); everything else can just be treated as a `u32` since `/dev/di` only passes the value on to the emulated title.

I also did some tests and found that the alignment on `DIMAR` and `DILENGTH` is enforced; bits 0-4 are always zero regardless of what's written (even if no command is triggered).  However, my testing found that bits 26-31 were kept for both, while dolphin previously was clearing them for `DIMAR`.